controlpanel/frontend/views/dashboard.py

```diff
@@ -0,0 +1,303 @@
+    success_url = reverse_lazy("list-dashboards")
+    allowed_methods = ["POST"]
+
+    def form_valid(self, form):
+        dashboard = self.get_object()
+        dashboard.delete()
+        messages.success(self.request, f"Successfully deleted {dashboard.name} dashboard")
+        return HttpResponseRedirect(self.get_success_url())
+
+
+@method_decorator(feature_flag_required("register_dashboard"), name="dispatch")
+class UpdateDashboard(
+    OIDCLoginRequiredMixin,
+    PermissionRequiredMixin,
+    SingleObjectMixin,
+    RedirectView,
+):
+    http_method_names = ["post"]
+    model = Dashboard
+
+    def get_redirect_url(self, *args, **kwargs):
+        return reverse_lazy("manage-dashboard", kwargs={"pk": kwargs["pk"]})
+
+    def post(self, request, *args, **kwargs):
+        self.perform_update(**kwargs)
+        return super().post(request, *args, **kwargs)
+
+
+@method_decorator(feature_flag_required("register_dashboard"), name="dispatch")
+class AddDashboardAdmin(UpdateDashboard):
+    model = Dashboard
+    permission_required = "api.add_dashboard_admin"
+
+    def perform_update(self, **kwargs):
+        dashboard = self.get_object()
+        user = get_object_or_404(User, pk=self.request.POST["user_id"])
+
+        dashboard.admins.add(user)
+        messages.success(self.request, f"Granted admin access to {user.name}")
+
+
+@method_decorator(feature_flag_required("register_dashboard"), name="dispatch")
+class RevokeDashboardAdmin(UpdateDashboard):
+    model = Dashboard
+    permission_required = "api.revoke_dashboard_admin"
+
+    def perform_update(self, **kwargs):
+        dashboard = self.get_object()
+        user = get_object_or_404(User, pk=kwargs["user_id"])
+
+        dashboard.admins.remove(user)
+        messages.success(self.request, f"Removed admin access from {user.name}")
+
+
+@method_decorator(feature_flag_required("register_dashboard"), name="dispatch")
+class DashboardCustomers(OIDCLoginRequiredMixin, PermissionRequiredMixin, DetailView):
+    context_object_name = "dashboard"
+    model = Dashboard
+    permission_required = "api.retrieve_dashboard"
+    template_name = "dashboard-customers-list.html"
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        dashboard = self.get_object()
+
+        customers = dashboard.viewers.all()
+        paginator = Paginator(customers, 25)
+
+        context["errors"] = self.request.session.pop("customer_form_errors", None)
+        context["page_no"] = page_no = self.kwargs.get("page_no")
+        context["paginator"] = paginator
+        context["elided"] = paginator.get_elided_page_range(page_no)
+        context["customers"] = customers
+        context["remove_customer_form"] = RemoveCustomerByEmailForm()
+        return context
+
+
+@method_decorator(feature_flag_required("register_dashboard"), name="dispatch")
+class AddDashboardCustomers(OIDCLoginRequiredMixin, PermissionRequiredMixin, UpdateView):
```

Comment:
`UpdateView` isnt the right class to use here, as it your form is not a modelform. So instead go for:
```suggestion
class AddDashboardCustomers(OIDCLoginRequiredMixin, PermissionRequiredMixin, SingleObjectMixin, FormView):
```
This means you still get [`get_object`](https://ccbv.co.uk/projects/Django/5.0/django.views.generic.detail/SingleObjectMixin/#get_object) from the `SingleObjectMixin` class.

And you still get all your form methods from the [`FormView` class](https://ccbv.co.uk/projects/Django/5.0/django.views.generic.edit/FormView/). Also means you can remove the hacky `get_form_kwargs` redefinition.